### ISSUE TITLE
Symlink following logic in src() seems wrong to me

### ISSUE THREAD

**Issue by xunnanxu** [gulpjs/vinyl-fs#132]
Basically, the generated vinyl file base should be what's set to to the symlink itself:

Say the src directory structure is like:
```
|-- foo
    |-- bar -> baz
    |-- baz
        |-- qux.txt
```
the symlink should produce: ./target/foo/bar at the end with qux.txt in it if dest is set to target. In practice it doesn't.

The reason is the file base should always be set to ./foo for anything linked by bar whereas the original recursive logic would produce foo/baz.

Such incorrect logic would yield a flattened file structure and empty directories at the end:

```
pipe(gulp.dest('target'))

the original logic would generate:

|-- target
    |-- foo
    |-- baz (empty)
    |-- qux.txt

whereas the fixed logic generates:

|-- target
    |-- foo
        |-- bar
            |-- qux.txt
```

I've opened a PR here: https://github.com/gulpjs/vinyl-fs/pull/131
Please check it out.

**Comment by phated** [comment]
Since this is a philosophical discussion, I'd like @contra and maybe even @piranna to weigh in on this.  Please take a look at the tests too, as it makes `isSymbolicLink()` return false.

**Comment by piranna** [comment]
I think `dest` should generate an exact clone of the file tree hierarchy that `src` has, without following symlinks at all. It said, `target` folder should have a `foo/bar` symlink pointing to `foo/baz` instead of replace the `foo/bar` symlink for a folder.

**Comment by xunnanxu** [comment]
@piranna Is what you said already a supported use case by using followSymlinks = false in src()? If followSymlinks is false (default), then we should follow what copy -RL does which turns symlinks into actual files and I believe that's what "follow symlinks" really should mean.

**Comment by piranna** [comment]
> Is what you said already a supported use case by using followSymlinks = false in src()?

Yes, seems it's already supported.

**Comment by xunnanxu** [comment]
Interestingly with this clearly being a bug and fix already provided, this issue is still open after three months and no one wants to take a look at it seriously.

**Comment by contra** [comment]
@xunnanxu If you'd like a quicker resolution feel free to send a PR with your ideal changes and we can review.

**Comment by GregDracoulis** [comment]
I'm running into this issue too. On any symlinked directories that I'm looking to recursively copy, any nested folders end up empty.

I'm using the `symlink` method as a temporary workaround until this is fixed. This might help a few people who just need to "copy" a whole folder structure recursively, where it doesn't matter if the files actually live in the target directory permanently (like for a dev server).

```JS
var vfs = require('vinyl-fs');

gulp.task('linkFolder', function() {
  return vfs.src(["src/myFolder/**/*"], {"followSymLinks": false})
    .pipe(vfs.symlink("build/dev/myFolder"));
});
```

On a separate note, it's really strange that if you point directly at the symlinked folder, it copies everything fine. To follow the example above:
```
|-- foo
    |-- bar -> baz
    |-- baz
        |-- qux.txt
```
if you use `["foo/**/*"]` you get the undesired behavior. However, if you point vinyl-fs at `["foo/bar/**/*"]` directly, it behaves as expected.

**Comment by phated** [comment]
@GregDracoulis ref https://github.com/gulpjs/vinyl-fs/pull/131#issuecomment-191991690

**Comment by floscher** [comment]
I have a similar problem, but only with Gulp 4.0.0-alpha.2, with Gulp 3.9.1 everything is fine. A minimal repository showing the issue is [floscher/gulp4-symlinks](//github.com/floscher/gulp4-symlinks).
floscher/gulp4-symlinks@7868019943b1a99c0efb57239381a4c113a4926b uses Gulp 3.9.1 and does not have the issue.
floscher/gulp4-symlinks@1d76800becd1fed84c566b580daa0e965f0b6819 uses Gulp 4.0.0-alpha.2 and  has the issue.


My directory structure is as follows:
```shell
├─┬ src
│ └─┬ foo
│   └─ baz → ../../vendor/bar
└─┬ vendor
  └─ bar   # a text file
```

And I am simply copying the files from `src` to `dest` with the following task:
```js
gulp.task('default',  () => {
  return gulp.src('src/**/*')
    .pipe(gulp.dest('dest'))
}}
```

Gulp 3.9.1 gives me:
```shell
└─┬ dest
  └─┬ foo
    └─ baz   # a copy of the file vendor/bar
```

But Gulp 4.0.0-alpha.2 gives me:
```shell
└─┬ dest
  ├── foo   # an empty directory
  └── bar   # a copy of the file vendor/bar
```

I consider the behaviour of Gulp 3.9.1 the correct one in this regard, because I use symlinks to change the directory structure of vendor-code. But when Gulp puts all symlinked files directly to the base-directory, my desired directory structure gets lost.
`followSymlinks`=true works, but only if I don't pipe the source through e.g. uglify() or sass().

Is this related to the issues described above? I first wanted to ask here instead of opening a separate issue, because it seems related.

**Comment by floscher** [comment]
I narrowed down the source of my issue: It was introduced between [gulpjs/vinyl-fs@`v0.3.13`](//github.com/gulpjs/vinyl-fs/tree/v0.1.13) and [gulpjs/vinyl-fs@`v1.0.0`](//github.com/gulpjs/vinyl-fs/tree/v1.0.0). That explains why it appears in Gulp 4 and not in Gulp 3.

**Comment by phated** [comment]
@floscher a ton was done between those versions (hence the breaking version bump) so that doesn't really help much.  These's an open PR related to this I'm waiting for follow up on.

**Comment by floscher** [comment]
@phated True, that info was not really precise. I'll investigate further where exactly the problem lies.

In the meantime I tested #131 and it does *not* fix my problem. Symbolic links on files still appear in the root of the destination folder using the name of the symbolic links target, not the name of the symbolic link itself.

Is it ok, that I discuss this issue here, or should I open another issue?

**Comment by floscher** [comment]
@phated True, that info was not very precise. But in the meantime I tried out PR #131 with my demo-repository and it fixes my problem.
Does this fix have any chance of being incorporated in the final Gulp 4? I'm not really familiar with the release plan of Gulp and how much time is left until v4 is released.

**Comment by xunnanxu** [comment]
@piranna @phated @contra It's been a while since the PR was updated. Could you guys take a look when you have a chance?

**Comment by phated** [comment]
@xunnanxu reviewed and commented.

@floscher yep, it will be in the gulp 4 release and in vinyl-fs sooner (which will be available through the gulp 4 branch)

**Comment by phated** [comment]
Published as 2.4.3 - many thanks to @xunnanxu
